src/main/java/duke/util/TaskList.java

```diff
@@ -0,0 +1,108 @@
+package duke.util;
+
+import duke.exception.DukeException;
+import duke.task.Task;
+
+import java.util.ArrayList;
+
+/**
+ * Represents a list of tasks defined by the user.
+ */
+public class TaskList {
+
+    private ArrayList<Task> taskList;
+
+    /**
+     * Initialises a new TaskList with no tasks.
+     */
+    public TaskList() {
+        taskList = new ArrayList<>();
+    }
+
+    /**
+     * Constructs a TaskList with preloaded tasks.
+     * @param taskList ArrayList of Task objects to be loaded
+     */
+    public TaskList(ArrayList<Task> taskList) {
+        this.taskList = taskList;
+    }
+
+    /**
+     * Deletes a Task.
+     * @param deletedItemNo index of the task
+     * @param ui ui object to print messages
+     * @throws DukeException if task is not found
+     */
+    public void delete(int deletedItemNo, Ui ui) throws DukeException {
+        try {
+            Task itemRemoved = taskList.remove(deletedItemNo - 1);
+            int numOfTask = taskList.size();
+            ui.showDeleted(itemRemoved, numOfTask);
+        } catch (Exception e) {
+            throw new DukeException("    No more tasks to delete!");
+        }
+    }
+
+    /**
+     * Adds a Task.
+     * @param task task object to be added
+     * @param ui ui object to print messages
+     */
+    public void add(Task task, Ui ui) {
+        taskList.add(task);
+        int numOfTask = taskList.size();
+        ui.showAdded(task, numOfTask);
+    }
+
+    /**
+     * Marks the task as done.
+     * @param itemNo index of the task to be marked
+     * @param ui ui object to print messages
+     */
+    public void done(int itemNo, Ui ui) {
+        Task task = taskList.get(itemNo - 1);
+        task.markAsDone();
+        ui.showDone(task);
+    }
+
+    /**
+     * Initiates writing of data into file.
+     * @param storage storage object to write data into
+     * @throws DukeException if file not found
+     */
+    public void writeTo(Storage storage) throws DukeException {
+        storage.writeToFile(taskList);
+    }
+
+    /**
+     * Finds the task(s) with matching descriptions.
+     * @param taskToBeFound search term
+     * @return TaskList object containing list of relevant search results
+     * @throws DukeException if task is not found
+     */
+    public TaskList find(String taskToBeFound) throws DukeException {
+        ArrayList<Task> searchResults = new ArrayList<>();
+        for (Task task : taskList) {
+            if (task.getDescription().contains(taskToBeFound)) {
+                searchResults.add(task);
+            }
+        }
+        if (searchResults.isEmpty()) {
+            throw new DukeException("     File not found! :(");
```

Comment:
Hmmm, since the extra spaces at the start are a specific UI requirement, would it be better to somehow add the indentation with a method at the `Ui` or higher class level instead?